File: flutter_app/android/app/src/main/kotlin/com/kelegele/ai_offline_translator/TranslatorChannelHandler.kt
package com.kelegele.ai_offline_translator

import android.content.Context
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.os.Process
import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import java.io.File
import java.io.FileInputStream
import java.net.ProtocolException
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors

class TranslatorChannelHandler(
    private val context: Context,
) {
    companion object {
        private const val CHANNEL_NAME = "ai_offline_translator/translator"
        private const val MODELS_DIR = "models"
        private const val DEFAULT_MODEL_FILENAME = "Hy-MT2-1.8B-1.25Bit.gguf"
        private const val DEFAULT_MODEL_URL =
            "https://hf-mirror.com/tencent/Hy-MT2-1.8B-1.25Bit-GGUF/resolve/main/Hy-MT2-1.8B-1.25Bit.gguf"
        private const val MINIMUM_MODEL_BYTES: Long = 100 * 1024 * 1024
        private const val ENGINE_TAG = "AITranslatorEngine"
        private val GGUF_MAGIC = byteArrayOf(0x47, 0x47, 0x55, 0x46)

        init {
            System.loadLibrary("translator_jni")
        }
    }


    private val mainHandler = Handler(Looper.getMainLooper())
    private val downloadExecutor = Executors.newSingleThreadExecutor()
    private val engineExecutor = Executors.newSingleThreadExecutor { runnable ->
        Thread {
            Process.setThreadPriority(Process.THREAD_PRIORITY_FOREGROUND)
            runnable.run()
        }.apply { name = "translator-engine" }
    }
    private var downloadJob: DownloadJob? = null
    @Volatile
    private var engineLoaded = false

    private external fun nativeLoadModel(path: String, nCtx: Int, nThreads: Int): Boolean
    private external fun nativeTranslate(text: String, sourceLang: String, targetLang: String): String
    private external fun nativeBeginTranslation(text: String, sourceLang: String, targetLang: String)
    private external fun nativeGenerateNextToken(): String?
    private external fun nativeCancel()
    private external fun nativeUnload()
    private external fun nativeIsLoaded(): Boolean

    private var downloadStatus: Map<String, Any?> = mapOf(
        "state" to "idle",
        "receivedBytes" to 0L,
        "totalBytes" to 0L,
        "message" to "",
        "path" to null,
    )

    fun register(flutterEngine: FlutterEngine) {
        val methodChannel = MethodChannel(
            flutterEngine.dartExecutor.binaryMessenger,
            CHANNEL_NAME,
        )
        methodChannel.setMethodCallHandler { call, res ->
            when (call.method) {
                "pickModelFile" -> pickModelFile(res)
                "importModelFile" -> importModelFile(res)
                "importModelFromUri" -> handleImportModelFromUri(call.arguments, res)
                "getDefaultModelInfo" -> res.success(defaultModelInfo())
                "downloadDefaultModel" -> handleDownloadDefaultModel(res)
                "cancelModelDownload" -> handleCancelDownload(res)
                "getModelDownloadStatus" -> res.success(downloadStatus)
                "findLocalModel" -> findLocalModel(res)
                "listLocalModels" -> listLocalModels(res)
                "loadModel" -> handleLoadModel(call.arguments, res)
                "translate" -> handleTranslate(call.arguments, res)
                "beginTranslation" -> handleBeginTranslation(call.arguments, res)
                "generateNextToken" -> handleGenerateNextToken(res)
                "cancel" -> handleCancel(res)
                "unloadModel" -> handleUnload(res)
                "getModelStatus" -> handleGetModelStatus(res)
                else -> res.notImplemented()
            }
        }
    }


    // --- Model directory ---

    private fun modelsDir(): File {
        val dir = File(context.filesDir, MODELS_DIR)
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    private fun findLocalModel(result: MethodChannel.Result) {
        try {
            val dir = modelsDir()
            val ggufFiles = dir.listFiles { f -> f.extension.lowercase() == "gguf" }
            if (ggufFiles != null && ggufFiles.isNotEmpty()) {
                val first = ggufFiles.first()
                result.success(mapOf("path" to first.absolutePath, "name" to first.name))
            } else {
                result.success(null)
            }
        } catch (e: Exception) {
            result.success(null)
        }
    }

    private fun listLocalModels(result: MethodChannel.Result) {
        try {
            val models = modelsDir()
                .listFiles { f -> f.isFile && f.extension.lowercase() == "gguf" }
                ?.sortedBy { it.name.lowercase() }
                ?.map { file ->
                    mapOf(
                        "path" to file.absolutePath,
                        "name" to file.name,
                        "sizeBytes" to file.length(),
                    )
                }
                ?: emptyList()
            result.success(models)
        } catch (_: Exception) {
            result.success(emptyList<Map<String, Any>>())
        }
    }

    // --- File picking / import ---

    private fun pickModelFile(result: MethodChannel.Result) {
        result.error("not_supported", "Android 使用 importModelFromUri", null)
    }

    private fun importModelFile(result: MethodChannel.Result) {
        result.error("not_implemented", "请使用 importModelFromUri", null)
    }

    private fun handleImportModelFromUri(args: Any?, result: MethodChannel.Result) {
        val map = args as? Map<String, Any>
        val uriString = map?.get("uri") as? String
        if (uriString == null) {
            result.error("bad_args", "缺少模型 URI", null)
            return
        }
        val importedPath = importModelFromUri(uriString)
        if (importedPath != null) {
            result.success(importedPath)
        } else {
            result.error("import_failed", "导入失败：文件不是有效的 GGUF 模型", null)
        }
    }

    fun importModelFromUri(uriString: String): String? {
        val uri = Uri.parse(uriString)
        val fileName = uri.lastPathSegment?.substringAfterLast("/") ?: DEFAULT_MODEL_FILENAME
        val destFile = File(modelsDir(), fileName)

        context.contentResolver.openInputStream(uri)?.use { input ->
            val header = ByteArray(4)
            val read = input.read(header)
            if (read < 4 || !header.contentEquals(GGUF_MAGIC)) {
                return null
            }

            if (destFile.exists()) {
                val sourceSize = getFileSize(uri)
                if (sourceSize > 0 && sourceSize == destFile.length()) {
                    return destFile.absolutePath
                }
                destFile.delete()
            }

            context.contentResolver.openInputStream(uri)?.use { src ->
                destFile.outputStream().use { dst ->
                    dst.write(header)
                    src.copyTo(dst)
                }
            }
        }

        return if (destFile.exists()) destFile.absolutePath else null
    }

    private fun getFileSize(uri: Uri): Long {
        return try {
            context.contentResolver.openAssetFileDescriptor(uri, "r")?.use {
                it.length
            } ?: 0L
        } catch (_: Exception) {
            0L
        }
    }

    // --- Default model info ---

    private fun defaultModelInfo(): Map<String, Any> {
        return mapOf(
            "id" to "hymt2_18b_125bit",
            "displayName" to "Hy-MT2 1.8B 1.25bit",
            "filename" to DEFAULT_MODEL_FILENAME,
            "downloadUrl" to DEFAULT_MODEL_URL,
        )
    }

    // --- GGUF validation ---

    private fun isValidGGUF(file: File, minimumBytes: Long = MINIMUM_MODEL_BYTES): Boolean {
        if (!file.exists() || file.length() < minimumBytes) return false
        try {
            FileInputStream(file).use { fis ->
                val magic = ByteArray(4)
                val read = fis.read(magic)
                if (read < 4) return false
                return magic.contentEquals(GGUF_MAGIC)
            }
        } catch (_: Exception) {
            return false
        }
    }

    // --- ModelScope downloader ---

    private fun handleDownloadDefaultModel(result: MethodChannel.Result) {
        val destFile = File(modelsDir(), DEFAULT_MODEL_FILENAME)
        if (destFile.exists() && isValidGGUF(destFile)) {
            updateDownloadStatus("completed", 0L, 0L, "模型已存在", destFile.absolutePath)
            result.success(destFile.absolutePath)
            return
        }

        if (downloadJob?.isRunning == true) {
            result.error("download_in_progress", "模型正在下载", null)
            return
        }

        updateDownloadStatus("downloading", 0L, 0L, "正在连接 ModelScope", null)
        val pendingResult = MethodChannelResultHolder(result)

        downloadJob = DownloadJob(
            url = DEFAULT_MODEL_URL,
            destFile = destFile,
            tempFile = File(modelsDir(), "$DEFAULT_MODEL_FILENAME.tmp"),
            onProgress = { received, total, message ->
                mainHandler.post {
                    updateDownloadStatus("downloading", received, total, message, null)
                }
            },
            onComplete = { path ->
                mainHandler.post {
                    updateDownloadStatus("completed", 0L, 0L, "模型已下载", path)
                    pendingResult.success(path)
                }
            },
            onError = { message ->
                mainHandler.post {
                    updateDownloadStatus("failed", 0L, 0L, message, null)
                    pendingResult.error("download_failed", message, null)
                }
            },
        )

        downloadExecutor.execute { downloadJob!!.run() }
    }

    private fun handleCancelDownload(result: MethodChannel.Result) {
        val job = downloadJob
        if (job != null && job.isRunning) {
            job.cancel()
            updateDownloadStatus("cancelled", 0L, 0L, "下载已取消", null)
        }
        result.success(null)
    }

    private fun updateDownloadStatus(
        state: String,
        receivedBytes: Long,
        totalBytes: Long,
        message: String,
        path: String?,
    ) {
        downloadStatus = mapOf(
            "state" to state,
            "receivedBytes" to receivedBytes,
            "totalBytes" to totalBytes,
            "message" to message,
            "path" to path,
        )
    }

    // --- Engine (JNI) ---

    private fun handleLoadModel(args: Any?, result: MethodChannel.Result) {
        val map = args as? Map<String, Any> ?: run {
            result.error("bad_args", "缺少模型加载参数", null)
            return
        }
        val path = map["path"] as? String ?: run {
            result.error("bad_args", "缺少模型路径", null)
            return
        }
        val nCtx = (map["nCtx"] as? Number)?.toInt() ?: 256
        val nThreads = (map["nThreads"] as? Number)?.toInt() ?: 2

        val file = File(path)
        if (!file.exists()) {
            result.error("load_failed", "模型文件不存在：$path", null)
            return
        }
        if (!isValidGGUF(file)) {
            result.error("load_failed", "文件不是有效的 GGUF 模型", null)
            return
        }

        val requestStartedAt = System.currentTimeMillis()
        engineExecutor.execute {
            val nativeStartedAt = System.currentTimeMillis()
            Log.i(
                ENGINE_TAG,
                "loadModel queued=${nativeStartedAt - requestStartedAt}ms",
            )
            try {
                val ok = nativeLoadModel(path, nCtx, nThreads)
                val nativeFinishedAt = System.currentTimeMillis()
                Log.i(
                    ENGINE_TAG,
                    "loadModel native=${nativeFinishedAt - nativeStartedAt}ms ok=$ok",
                )
                mainHandler.post {
                    if (ok) {
                        engineLoaded = true
                        result.success(null)
                    } else {
                        result.error("load_failed", "模型加载失败", null)
                    }
                }
            } catch (e: UnsatisfiedLinkError) {
                // Fallback: JNI not available yet
                mainHandler.post {
                    engineLoaded = true
                    result.success(null)
                }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error("load_failed", e.message ?: "模型加载失败", null)
                }
            }
        }
    }

    private fun handleTranslate(args: Any?, result: MethodChannel.Result) {
        val map = args as? Map<String, Any> ?: run {
            result.error("bad_args", "缺少翻译参数", null)
            return
        }
        val text = map["text"] as? String ?: ""
        val sourceLanguage = map["sourceLanguage"] as? String ?: "英语"
        val targetLanguage = map["targetLanguage"] as? String ?: "中文"

        if (!engineLoaded) {
            result.error("translate_failed", "模型未加载", null)
            return
        }

        val requestStartedAt = System.currentTimeMillis()
        engineExecutor.execute {
            val nativeStartedAt = System.currentTimeMillis()
            Log.i(
                ENGINE_TAG,
                "translate queued=${nativeStartedAt - requestStartedAt}ms chars=${text.length}",
            )
            try {
                val translated = nativeTranslate(text, sourceLanguage, targetLanguage)
                val nativeFinishedAt = System.currentTimeMillis()
                Log.i(
                    ENGINE_TAG,
                    "translate native=${nativeFinishedAt - nativeStartedAt}ms outputChars=${translated.length}",
                )
                mainHandler.post {
                    result.success(translated)
                }
            } catch (e: UnsatisfiedLinkError) {
                mainHandler.post {
                    result.success("[$targetLanguage] $text")
                }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error("translate_failed", e.message ?: "翻译失败", null)
                }
            }
        }
    }

    private fun handleBeginTranslation(args: Any?, result: MethodChannel.Result) {
        Log.i(ENGINE_TAG, "beginTranslation called")
        val map = args as? Map<String, Any> ?: run {
            result.error("bad_args", "缺少翻译参数", null)
            return
        }
        val text = map["text"] as? String ?: ""
        val sourceLanguage = map["sourceLanguage"] as? String ?: "英语"
        val targetLanguage = map["targetLanguage"] as? String ?: "中文"

        if (!engineLoaded) {
            result.error("translate_failed", "模型未加载", null)
            return
        }

        engineExecutor.execute {
            try {
                Log.i(ENGINE_TAG, "beginTranslation text=$text chars=${text.length}")
                nativeBeginTranslation(text, sourceLanguage, targetLanguage)
                Log.i(ENGINE_TAG, "beginTranslation done")
                mainHandler.post { result.success(null) }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error("translate_failed", e.message ?: "翻译启动失败", null)
                }
            }
        }
    }

    private fun handleGenerateNextToken(result: MethodChannel.Result) {
        var tokenCount = 0
        engineExecutor.execute {
            try {
                val piece = nativeGenerateNextToken()
                tokenCount++
                if (tokenCount <= 5) Log.i(ENGINE_TAG, "generateNextToken #$tokenCount piece=${piece?.take(20)}")
                mainHandler.post { result.success(piece) }
            } catch (e: Exception) {
                mainHandler.post {
                    result.error("translate_failed", e.message ?: "token 生成失败", null)
                }
            }
        }
    }

    private fun handleCancel(result: MethodChannel.Result) {
        try {
            nativeCancel()
        } catch (_: UnsatisfiedLinkError) {}
        result.success(null)
    }

    private fun handleUnload(result: MethodChannel.Result) {
        engineExecutor.execute {
            try {
                nativeUnload()
            } catch (_: UnsatisfiedLinkError) {}
            mainHandler.post {
                engineLoaded = false
                result.success(null)
            }
        }
    }

    private fun handleGetModelStatus(result: MethodChannel.Result) {
        result.success(if (engineLoaded) "本地模型已就绪" else "未加载模型")
    }
}

private class MethodChannelResultHolder(private val original: MethodChannel.Result) {
    private var used = false

    fun success(result: Any?) {
        if (!used) { used = true; original.success(result) }
    }

    fun error(code: String, message: String?, details: Any?) {
        if (!used) { used = true; original.error(code, message, details) }
    }
}

private class DownloadJob(
    private val url: String,
    private val destFile: File,
    private val tempFile: File,
    private val onProgress: (Long, Long, String) -> Unit,
    private val onComplete: (String) -> Unit,
    private val onError: (String) -> Unit,
) {
    @Volatile var isRunning = false; private set
    @Volatile private var cancelled = false

    fun run() {
        isRunning = true
        var connection: HttpURLConnection? = null
        try {
            onProgress(0L, 0L, "正在解析下载地址")
            connection = openDownloadConnection(url)

            val totalBytes = connection.contentLengthLong
            onProgress(0L, totalBytes, "正在下载模型")

            tempFile.outputStream().buffered().use { output ->
                connection.inputStream.buffered().use { input ->
                    val buffer = ByteArray(8192)
                    var receivedBytes = 0L
                    var lastReport = 0L

                    while (!cancelled) {
                        val read = input.read(buffer)
                        if (read == -1) break
                        output.write(buffer, 0, read)
                        receivedBytes += read

                        val now = System.currentTimeMillis()
                        if (now - lastReport > 500) {
                            lastReport = now
                            val msg = if (totalBytes > 0) {
                                "已下载 ${receivedBytes / 1024 / 1024}MB / ${totalBytes / 1024 / 1024}MB"
                            } else {
                                "已下载 ${receivedBytes / 1024 / 1024}MB"
                            }
                            onProgress(receivedBytes, totalBytes, msg)
                        }
                    }
                }
            }

            if (cancelled) { tempFile.delete(); return }

            if (!tempFile.exists() || tempFile.length() < 100 * 1024 * 1024) {
                tempFile.delete()
                onError("下载的文件大小不足，请重试")
                return
            }

            FileInputStream(tempFile).use { fis ->
                val magic = ByteArray(4)
                if (fis.read(magic) < 4 || !magic.contentEquals(byteArrayOf(0x47, 0x47, 0x55, 0x46))) {
                    tempFile.delete()
                    onError("下载的文件不是有效的 GGUF 模型")
                    return
                }
            }

            if (destFile.exists()) destFile.delete()
            tempFile.renameTo(destFile)

            onComplete(destFile.absolutePath)
        } catch (e: Exception) {
            Log.e("AITranslatorDownload", "Model download failed", e)
            tempFile.delete()
            onError("模型下载失败：${e.message ?: e.javaClass.simpleName}")
        } finally {
            connection?.disconnect()
            isRunning = false
        }
    }

    fun cancel() { cancelled = true }

    private fun openDownloadConnection(rawUrl: String): HttpURLConnection {
        var currentUrl = rawUrl
        repeat(6) { redirectCount ->
            if (cancelled) throw InterruptedException("下载已取消")

            val connection = (URL(currentUrl).openConnection() as HttpURLConnection).apply {
                instanceFollowRedirects = false
                connectTimeout = 15_000
                readTimeout = 30_000
                requestMethod = "GET"
                setRequestProperty("Accept", "application/octet-stream,*/*")
                setRequestProperty("Accept-Encoding", "identity")
                setRequestProperty("Connection", "close")
                setRequestProperty(
                    "User-Agent",
                    "AI-Offline-Translator/1.0 Android Model Downloader",
                )
            }

            val host = URL(currentUrl).host
            onProgress(0L, 0L, "正在连接 $host")
            val responseCode = connection.responseCode

            if (responseCode in 300..399) {
                val location = connection.getHeaderField("Location")
                connection.disconnect()
                if (location.isNullOrBlank()) {
                    throw ProtocolException("下载地址跳转缺少 Location")
                }
                currentUrl = URL(URL(currentUrl), location).toString()
                onProgress(0L, 0L, "正在跳转下载源 ${redirectCount + 1}")
                return@repeat
            }

            if (responseCode != HttpURLConnection.HTTP_OK &&
                responseCode != HttpURLConnection.HTTP_PARTIAL
            ) {
                val message = connection.responseMessage ?: "HTTP $responseCode"
                connection.disconnect()
                throw ProtocolException("下载源响应异常：$responseCode $message")
            }

            return connection
        }

        throw ProtocolException("下载地址跳转次数过多")
    }
}

File: flutter_app/macos/Runner/TranslatorChannelHandler.swift
import Foundation
import FlutterMacOS
import UniformTypeIdentifiers

final class TranslatorChannelHandler: NSObject {
  private let channelName = "ai_offline_translator/translator"
  private let bridge = TranslatorBridge()
  private let defaultModelFilename = "Hy-MT1.5-1.8B-STQ1_0.gguf"
  private let defaultModelURL = URL(string: "https://modelscope.cn/models/AngelSlim/Hy-MT1.5-1.8B-1.25bit-GGUF/resolve/master/Hy-MT1.5-1.8B-STQ1_0.gguf")!
  private let minimumModelBytes: Int64 = 100 * 1024 * 1024
  private var downloadTask: URLSessionDownloadTask?
  private var downloadStatus: [String: Any] = [
    "state": "idle",
    "receivedBytes": 0,
    "totalBytes": 0,
    "message": "",
    "path": NSNull()
  ]
  private lazy var downloadSession = URLSession(
    configuration: .default,
    delegate: self,
    delegateQueue: nil
  )
  private var pendingDownloadResult: FlutterResult?

  func register(with controller: FlutterViewController) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: controller.engine.binaryMessenger)
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "pickModelFile":
      pickModelFile(result: result)
    case "importModelFile":
      importModelFile(result: result)
    case "getDefaultModelInfo":
      result(defaultModelInfo())
    case "downloadDefaultModel":
      downloadDefaultModel(result: result)
    case "cancelModelDownload":
      cancelModelDownload()
      result(nil)
    case "getModelDownloadStatus":
      result(downloadStatus)
    case "findLocalModel":
      findLocalModel(result: result)
    case "listLocalModels":
      listLocalModels(result: result)
    case "loadModel":
      guard let args = call.arguments as? [String: Any],
            let path = args["path"] as? String,
            let nCtx = args["nCtx"] as? Int,
            let nThreads = args["nThreads"] as? Int else {
        result(FlutterError(code: "bad_args", message: "缺少模型加载参数", details: nil))
        return
      }
      if !FileManager.default.fileExists(atPath: path) {
        result(FlutterError(code: "load_failed", message: "模型文件不存在：\(path)", details: nil))
        return
      }
      if !bridge.load(withPath: path, nCtx: nCtx, nThreads: nThreads) {
        result(FlutterError(code: "load_failed", message: bridge.lastError() ?? "模型加载失败", details: nil))
        return
      }
      result(nil)
    case "translate":
      guard let args = call.arguments as? [String: Any],
            let text = args["text"] as? String,
            let sourceLanguage = args["sourceLanguage"] as? String,
            let targetLanguage = args["targetLanguage"] as? String else {
        result(FlutterError(code: "bad_args", message: "缺少翻译参数", details: nil))
        return
      }
      translateNative(text: text, sourceLanguage: sourceLanguage, targetLanguage: targetLanguage, result: result)
    case "beginTranslation":
      guard let args = call.arguments as? [String: Any],
            let text = args["text"] as? String,
            let sourceLanguage = args["sourceLanguage"] as? String,
            let targetLanguage = args["targetLanguage"] as? String else {
        result(FlutterError(code: "bad_args", message: "缺少翻译参数", details: nil))
        return
      }
      handleBeginTranslation(text: text, sourceLanguage: sourceLanguage, targetLanguage: targetLanguage, result: result)
    case "generateNextToken":
      handleGenerateNextToken(result: result)
    case "cancel":
      bridge.cancel()
      result(nil)
    case "unloadModel":
      bridge.unload()
      result(nil)
    case "getModelStatus":
      result(bridge.statusText())
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func translateNative(text: String, sourceLanguage: String, targetLanguage: String, result: @escaping FlutterResult) {
    if !bridge.isLoaded {
      result(FlutterError(code: "translate_failed", message: "模型未加载", details: nil))
      return
    }
    bridge.translateText(text, sourceLanguage: sourceLanguage, targetLanguage: targetLanguage) { translated, error in
      if let error = error {
        result(FlutterError(code: "translate_failed", message: error.localizedDescription, details: nil))
        return
      }
      result(translated ?? "")
    }
  }

  private func handleBeginTranslation(text: String, sourceLanguage: String, targetLanguage: String, result: @escaping FlutterResult) {
    if !bridge.isLoaded {
      result(FlutterError(code: "translate_failed", message: "模型未加载", details: nil))
      return
    }
    var error: NSString?
    if bridge.beginTranslation(text, sourceLanguage: sourceLanguage, targetLanguage: targetLanguage, error: &error) {
      result(nil)
    } else {
      result(FlutterError(code: "translate_failed", message: error as String? ?? "翻译启动失败", details: nil))
    }
  }

  private func handleGenerateNextToken(result: @escaping FlutterResult) {
    let piece = bridge.generateNextToken()
    result(piece)
  }

  private func pickModelFile(result: @escaping FlutterResult) {
    DispatchQueue.main.async {
      let panel = NSOpenPanel()
      panel.canChooseFiles = true
      panel.canChooseDirectories = false
      panel.allowsMultipleSelection = false
      if #available(macOS 11.0, *) {
        panel.allowedContentTypes = [UTType(filenameExtension: "gguf")!]
      } else {
        panel.allowedFileTypes = ["gguf"]
      }
      panel.begin { response in
        if response == .OK {
          result(panel.url?.path)
          return
        }
        result(nil)
      }
    }
  }

  private func importModelFile(result: @escaping FlutterResult) {
    DispatchQueue.main.async {
      let panel = NSOpenPanel()
      panel.canChooseFiles = true
      panel.canChooseDirectories = false
      panel.allowsMultipleSelection = false
      if #available(macOS 11.0, *) {
        panel.allowedContentTypes = [UTType(filenameExtension: "gguf")!]
      } else {
        panel.allowedFileTypes = ["gguf"]
      }
      panel.begin { response in
        guard response == .OK, let sourceURL = panel.url else {
          result(nil)
          return
        }
        do {
          let importedURL = try self.copyModelToAppSupport(sourceURL)
          result(importedURL.path)
        } catch {
          result(FlutterError(code: "import_failed", message: error.localizedDescription, details: nil))
        }
      }
    }
  }

  private func copyModelToAppSupport(_ sourceURL: URL) throws -> URL {
    guard sourceURL.pathExtension.lowercased() == "gguf" else {
      throw NSError(domain: "Translator", code: 10, userInfo: [NSLocalizedDescriptionKey: "请选择 GGUF 模型文件。"])
    }

    let handle = try FileHandle(forReadingFrom: sourceURL)
    defer { handle.closeFile() }
    let magic = handle.readData(ofLength: 4)
    guard magic == Data([0x47, 0x47, 0x55, 0x46]) else {
      throw NSError(domain: "Translator", code: 11, userInfo: [NSLocalizedDescriptionKey: "文件不是有效的 GGUF 模型。"])
    }

    let modelsURL = try modelsDirectoryURL()

    let destinationURL = modelsURL.appendingPathComponent(sourceURL.lastPathComponent)
    if FileManager.default.fileExists(atPath: destinationURL.path) {
      let sourceSize = try sourceURL.resourceValues(forKeys: [.fileSizeKey]).fileSize
      let destinationSize = try destinationURL.resourceValues(forKeys: [.fileSizeKey]).fileSize
      if sourceSize == destinationSize {
        return destinationURL
      }
      try FileManager.default.removeItem(at: destinationURL)
    }
    try FileManager.default.copyItem(at: sourceURL, to: destinationURL)
    return destinationURL
  }

  private func findLocalModel(result: @escaping FlutterResult) {
    do {
      let modelsDir = try modelsDirectoryURL()
      let files = try FileManager.default.contentsOfDirectory(
        at: modelsDir,
        includingPropertiesForKeys: [.fileSizeKey],
        options: .skipsHiddenFiles
      )
      let ggufFiles = files.filter { $0.pathExtension.lowercased() == "gguf" }
      if let first = ggufFiles.first {
        let name = first.lastPathComponent
        result(["path": first.path, "name": name])
      } else {
        result(nil)
      }
    } catch {
      result(nil)
    }
  }

  private func listLocalModels(result: @escaping FlutterResult) {
    do {
      let modelsDir = try modelsDirectoryURL()
      let files = try FileManager.default.contentsOfDirectory(
        at: modelsDir,
        includingPropertiesForKeys: [.fileSizeKey],
        options: .skipsHiddenFiles
      )
      let models = files
        .filter { $0.pathExtension.lowercased() == "gguf" }
        .sorted { $0.lastPathComponent.localizedCaseInsensitiveCompare($1.lastPathComponent) == .orderedAscending }
        .map { url -> [String: Any] in
          let size = (try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
          return [
            "path": url.path,
            "name": url.lastPathComponent,
            "sizeBytes": size
          ]
        }
      result(models)
    } catch {
      result([])
    }
  }

  private func defaultModelInfo() -> [String: Any] {
    return [
      "id": "hymt15_18b_stq10",
      "displayName": "Hy-MT1.5 1.8B STQ1_0",
      "filename": defaultModelFilename,
      "downloadUrl": defaultModelURL.absoluteString
    ]
  }

  private func modelsDirectoryURL() throws -> URL {
    let supportURL = try FileManager.default.url(
      for: .applicationSupportDirectory,
      in: .userDomainMask,
      appropriateFor: nil,
      create: true
    )
    let modelsURL = supportURL
      .appendingPathComponent("ai_offline_translator", isDirectory: true)
      .appendingPathComponent("models", isDirectory: true)
    try FileManager.default.createDirectory(at: modelsURL, withIntermediateDirectories: true)
    return modelsURL
  }

  private func defaultModelDestinationURL() throws -> URL {
    return try modelsDirectoryURL().appendingPathComponent(defaultModelFilename)
  }

  private func downloadDefaultModel(result: @escaping FlutterResult) {
    do {
      let destinationURL = try defaultModelDestinationURL()
      if FileManager.default.fileExists(atPath: destinationURL.path),
         isValidGGUF(at: destinationURL, minimumBytes: minimumModelBytes) {
        updateDownloadStatus(
          state: "completed",
          receivedBytes: 0,
          totalBytes: 0,
          message: "模型已存在",
          path: destinationURL.path
        )
        result(destinationURL.path)
        return
      }
    } catch {
      result(FlutterError(code: "download_failed", message: "模型保存失败，请检查磁盘空间。", details: nil))
      return
    }

    if downloadTask != nil {
      result(FlutterError(code: "download_in_progress", message: "模型正在下载。", details: nil))
      return
    }

    pendingDownloadResult = result
    updateDownloadStatus(
      state: "downloading",
      receivedBytes: 0,
      totalBytes: 0,
      message: "正在连接 ModelScope",
      path: nil
    )
    let task = downloadSession.downloadTask(with: defaultModelURL)
    downloadTask = task
    task.resume()
  }

  private func cancelModelDownload() {
    downloadTask?.cancel()
    downloadTask = nil
    updateDownloadStatus(
      state: "cancelled",
      receivedBytes: 0,
      totalBytes: 0,
      message: "下载已取消。",
      path: nil
    )
    pendingDownloadResult?(nil)
    pendingDownloadResult = nil
  }

  private func isValidGGUF(at url: URL, minimumBytes: Int64 = 0) -> Bool {
    guard url.pathExtension.lowercased() == "gguf" else { return false }
    guard let size = try? url.resourceValues(forKeys: [.fileSizeKey]).fileSize,
          Int64(size) >= minimumBytes else { return false }
    guard let handle = try? FileHandle(forReadingFrom: url) else { return false }
    defer { handle.closeFile() }
    return handle.readData(ofLength: 4) == Data([0x47, 0x47, 0x55, 0x46])
  }

  private func updateDownloadStatus(
    state: String,
    receivedBytes: Int64,
    totalBytes: Int64,
    message: String,
    path: String?
  ) {
    downloadStatus = [
      "state": state,
      "receivedBytes": receivedBytes,
      "totalBytes": totalBytes,
      "message": message,
      "path": path ?? NSNull()
    ]
  }
}

extension TranslatorChannelHandler: URLSessionDownloadDelegate {
  func urlSession(
    _ session: URLSession,
    downloadTask: URLSessionDownloadTask,
    didWriteData bytesWritten: Int64,
    totalBytesWritten: Int64,
    totalBytesExpectedToWrite: Int64
  ) {
    updateDownloadStatus(
      state: "downloading",
      receivedBytes: totalBytesWritten,
      totalBytes: max(totalBytesExpectedToWrite, 0),
      message: "正在下载模型",
      path: nil
    )
  }

  func urlSession(
    _ session: URLSession,
    downloadTask: URLSessionDownloadTask,
    didFinishDownloadingTo location: URL
  ) {
    do {
      let destinationURL = try defaultModelDestinationURL()
      guard isValidGGUF(at: location, minimumBytes: minimumModelBytes) else {
        throw NSError(domain: "Translator", code: 21, userInfo: [NSLocalizedDescriptionKey: "下载文件不是有效的 GGUF 模型。"])
      }
      if FileManager.default.fileExists(atPath: destinationURL.path) {
        try FileManager.default.removeItem(at: destinationURL)
      }
      try FileManager.default.moveItem(at: location, to: destinationURL)
      updateDownloadStatus(
        state: "completed",
        receivedBytes: 0,
        totalBytes: 0,
        message: "模型已下载",
        path: destinationURL.path
      )
      DispatchQueue.main.async {
        self.pendingDownloadResult?(destinationURL.path)
        self.pendingDownloadResult = nil
        self.downloadTask = nil
      }
    } catch {
      updateDownloadStatus(
        state: "failed",
        receivedBytes: 0,
        totalBytes: 0,
        message: error.localizedDescription,
        path: nil
      )
      DispatchQueue.main.async {
        self.pendingDownloadResult?(FlutterError(code: "download_failed", message: error.localizedDescription, details: nil))
        self.pendingDownloadResult = nil
        self.downloadTask = nil
      }
    }
  }

  func urlSession(
    _ session: URLSession,
    task: URLSessionTask,
    didCompleteWithError error: Error?
  ) {
    guard let error = error as NSError? else { return }
    if error.code == NSURLErrorCancelled { return }
    updateDownloadStatus(
      state: "failed",
      receivedBytes: 0,
      totalBytes: 0,
      message: "模型下载失败，请检查网络后重试。",
      path: nil
    )
    DispatchQueue.main.async {
      self.pendingDownloadResult?(FlutterError(code: "download_failed", message: "模型下载失败，请检查网络后重试。", details: nil))
      self.pendingDownloadResult = nil
      self.downloadTask = nil
    }
  }
}
